File: examples/javascript/session_sample.ts
/**
 * Moss SDK - Session Sample
 *
 * A `SessionIndex` is a local, in-process index: documents are embedded and
 * queried entirely in memory with no cloud round-trip. At the end of a session
 * call `pushIndex()` to persist it to the cloud so another agent or device can
 * resume it via `loadIndex()`.
 *
 * Typical use-cases:
 *  - Indexing a live conversation turn-by-turn as it happens
 *  - Ephemeral per-user or per-request search spaces
 *  - Offline-first mobile apps that sync on demand
 *
 * Required Environment Variables:
 * - MOSS_PROJECT_ID: Your Moss project ID
 * - MOSS_PROJECT_KEY: Your Moss project key
 * - MOSS_INDEX_NAME: Name for the session index. Defaults to "session-demo".
 *
 * @example
 * ```bash
 * npm run session
 * ```
 */

import { MossClient, DocumentInfo } from "@moss-dev/moss";
import { config } from "dotenv";

config();

async function sessionSample(): Promise<void> {
  console.log("Moss SDK - Session Sample");
  console.log("=".repeat(40));

  const projectId = process.env.MOSS_PROJECT_ID;
  const projectKey = process.env.MOSS_PROJECT_KEY;
  const indexName = process.env.MOSS_INDEX_NAME ?? "session-demo";

  if (!projectId || !projectKey) {
    console.error("Error: Please set MOSS_PROJECT_ID and MOSS_PROJECT_KEY in .env");
    return;
  }

  const client = new MossClient(projectId, projectKey);

  try {
    // Open a session. If a cloud index with this name already exists it is
    // loaded into the session automatically; otherwise it starts empty.
    console.log(`\nOpening session '${indexName}'...`);
    const session = await client.session(indexName);
    console.log(`Session open (${session.docCount} pre-existing docs, model: ${session.modelId})`);

    // Add documents locally — embeddings are generated in Rust, no network call.
    console.log("\nAdding documents locally...");
    const turns: DocumentInfo[] = [
      { id: "turn-1", text: "Customer was charged twice for the March renewal." },
      { id: "turn-2", text: "Agent confirmed a refund for the duplicate charge." },
      { id: "turn-3", text: "Customer also asked to cancel auto-renew." },
    ];
    const { added, updated } = await session.addDocs(turns);
    console.log(`${added} added, ${updated} updated (${session.docCount} total in session)`);

    // Retrieve the documents we just added.
    console.log("\nRetrieving all session docs...");
    const docs = await session.getDocs();
    docs.forEach((doc) => {
      console.log(`  [${doc.id}] ${doc.text}`);
    });

    // Query the in-memory index — typically 1-10 ms with no network latency.
    console.log("\nQuerying the session...");
    const results = await session.query("what did the customer want refunded", { topK: 3 });
    console.log(`${results.docs.length} results:`);
    results.docs.forEach((doc) => {
      console.log(`  [${doc.id}] score=${doc.score.toFixed(3)}  ${doc.text}`);
    });

    // Push the session to the cloud so another process can resume it.
    console.log("\nPushing session to the cloud...");
    const pushed = await session.pushIndex();
    console.log(`Pushed ${pushed.docCount} docs (job ${pushed.jobId})`);

    console.log("\nDone! Another agent can now resume this session with: (Ctrl+C to exit)");
    console.log(`  const session = await client.session('${indexName}');`);
  } catch (error) {
    console.error(`Error: ${error}`);
  }
}

export { sessionSample };

if (require.main === module) {
  sessionSample().catch(console.error);
}

File: examples/python/session_sample.py
"""
Moss SDK Session Sample

A SessionIndex is a local, in-process index you read and write in real time with no
cloud round trip, then push to the cloud so another agent or device can resume it.
This is how Moss indexes a live conversation (e.g. voice/chat transcript turns).

Required Environment Variables:
- MOSS_PROJECT_ID: Your Moss project ID
- MOSS_PROJECT_KEY: Your Moss project key
- MOSS_INDEX_NAME: Name for the session index (created or resumed). Defaults to "session-demo".
"""

import asyncio
import os
from dotenv import load_dotenv
from moss import MossClient, DocumentInfo, QueryOptions, GetDocumentsOptions

# Load environment variables
load_dotenv()


async def session_sample():
    """Open a session, index documents locally, query in-memory, then push to the cloud."""
    print("=" * 40)
    print("Moss SDK - Session Sample")
    print("=" * 40)

    project_id = os.getenv("MOSS_PROJECT_ID")
    project_key = os.getenv("MOSS_PROJECT_KEY")
    index_name = os.getenv("MOSS_INDEX_NAME", "session-demo")

    if not project_id or not project_key:
        print("Error: Missing required environment variables!")
        print("Please set MOSS_PROJECT_ID and MOSS_PROJECT_KEY in .env file")
        return

    client = MossClient(project_id, project_key)

    try:
        # Open a session by name. If a cloud index with this name already exists it
        # auto-loads; otherwise the session starts empty. No cloud round trip on add/query.
        print(f"\nOpening session '{index_name}'...")
        session = await client.session(index_name=index_name)
        print(f"Session '{session.name}' open ({session.doc_count} existing docs)")

        # Add documents as they arrive (e.g. transcript turns). Embedded locally.
        print("\nAdding documents locally...")
        added, updated = await session.add_docs([
            DocumentInfo(id="turn-1", text="Customer was charged twice for the March renewal."),
            DocumentInfo(id="turn-2", text="Agent confirmed a refund for the duplicate charge."),
            DocumentInfo(id="turn-3", text="Customer also asked to cancel auto-renew."),
            DocumentInfo(id="turn-4", text="Agent placed a cancellation request for auto-renew."),
        ])
        print(f"{added} added, {updated} updated ({session.doc_count} total)")

        # Retrieve all documents currently in the session index.
        print("\nRetrieving all session documents...")
        all_docs = await session.get_docs()
        print(f"Found {len(all_docs)} docs:")
        for doc in all_docs:
            print(f"  [{doc.id}] {doc.text}")

        # Retrieve specific documents by ID.
        print("\nRetrieving specific documents by ID...")
        specific = await session.get_docs(GetDocumentsOptions(doc_ids=["turn-1", "turn-3"]))
        print(f"Fetched {len(specific)} docs:")
        for doc in specific:
            print(f"  [{doc.id}] {doc.text}")

        # Query the in-memory session (~1-10 ms, no network).
        print("\nQuerying the session...")
        results = await session.query("what did the customer want refunded", QueryOptions(top_k=3))
        for doc in results.docs:
            print(f"  [{doc.id}] {doc.score:.3f}  {doc.text}")

        # Delete a document that is no longer needed.
        print("\nDeleting 'turn-4' from session...")
        deleted_count = await session.delete_docs(["turn-4"])
        print(f"Deleted {deleted_count} doc(s) ({session.doc_count} remaining)")

        # Query again to confirm the deletion.
        print("\nQuerying after deletion...")
        results_after = await session.query("auto-renew cancellation", QueryOptions(top_k=3))
        for doc in results_after.docs:
            print(f"  [{doc.id}] {doc.score:.3f}  {doc.text}")

        # Push the session to the cloud so another agent or device can resume it.
        print("\nPushing session to the cloud...")
        pushed = await session.push_index()
        print(f"Pushed {pushed.doc_count} docs to '{pushed.index_name}' (job {pushed.job_id}, status: {pushed.status})")

        print("\nSample completed successfully!")

    except Exception as error:
        print(f"Error: {error}")
        print("Check your credentials in .env file")


# Export main function
__all__ = ["session_sample"]


# Run the sample
if __name__ == "__main__":
    asyncio.run(session_sample())
